File: Assets/Unity/Network/AppError.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class AppError: Exception {

    // MARK: - Property
    
    public ErrorKind kind;

    // MARK: - Initializer

    public AppError(ErrorKind kind) {
        this.kind = kind;
    }

    public AppError(string statusCode) {
        switch (statusCode) {
            case "401":
                this.kind = ErrorKind.unauthorized;
                break;

            case "402":
                this.kind = ErrorKind.badrequest;
                break;

            case "403":
                this.kind = ErrorKind.forbidden;
                break;

            case "404":
                this.kind = ErrorKind.notfound;
                break;

            case "500":
            case "501":
            case "502":
            case "503":
                this.kind = ErrorKind._internal;
                break;
        }
    }

    // MARK: - Builder

    public static void checkError(string statusCode) {
        if (statusCode.Length == 3 && int.Parse(statusCode) >= 400) {
            throw new AppError(statusCode);
        }
    }

    public static void checkError(int statusCode) {
        if (statusCode >= 400) {
            throw new AppError("" + statusCode);
        }
    }
}

public enum ErrorKind {
        unauthorized,
        badrequest,
        forbidden,
        notfound,
        _internal,
        network,
    }

File: Assets/Plugin/iOS/Swift/Networking/AppError.swift
import Moya

enum NetworkingError: Error {
    case unauthorized(message: String?)
    case badrequest(message: String?)
    case forbidden(message: String?)
    case notfound(message: String?)
    case `internal`(message: String?)
    case network
    case unknown

    struct Response: Decodable {
        let message: String
    }

    var statusCode: Int {
        switch self {
        case .badrequest:
            return 400
        case .unauthorized:
            return 401
        case .forbidden:
            return 403
        case .notfound:
            return 404
        case .internal:
            return 500
        case .network:
            return 600
        case .unknown:
            return 700
        }
    }

    // MARK: - Initializer

    init(error: Error) {
        guard let moyaError = error as? MoyaError else {
            self = .unknown
            return
        }

        let message = try? JSONDecoder().decode(Response.self, from: moyaError.response?.data ?? Data()).message

        switch moyaError.response?.statusCode {
        case 400:
            self = .badrequest(message: message)
        case 401:
            self = .unauthorized(message: message)
        case 403:
            self = .forbidden(message: message)
        case 404:
            self = .notfound(message: message)
        case 500, 501, 502, 503, 504, 505:
            self = .internal(message: message)
        default:
            self = .network
        }
    }
}
